crates/aiken-lang/src/tipo/expr.rs

```diff
@@ -1707,13 +1711,144 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
+            //
+            // Now, whether this leads to an invalid call usage, that's not *our* immediate
+            // problem.
+            UntypedExpr::Fn {
+                location: call_location,
+                fn_style,
+                ref arguments,
+                ref return_annotation,
+                ..
+            } => {
+                let return_annotation = return_annotation.clone();
+
+                let arguments = arguments.iter().skip(1).cloned().collect::<Vec<_>>();
+
+                let call = UntypedExpr::Call {
+                    location: call_location,
+                    fun: value,
+                    arguments: vec![CallArg {
+                        location: assign_location,
+                        label: None,
+                        value: UntypedExpr::lambda(name, continuation, call_location),
+                    }],
+                };
+
+                if arguments.is_empty() {
+                    call
+                } else {
+                    UntypedExpr::Fn {
+                        location: call_location,
+                        fn_style,
+                        arguments,
+                        body: call.into(),
+                        return_annotation,
+                    }
+                }
+            }
+
+            // Similarly to function captures, if we have any other expression we simply call it
+            // with our continuation. If the expression isn't callable? No problem, the
+            // type-checker will catch that eventually in exactly the same way as if the code was
+            // written like that to begin with.
+            _ => UntypedExpr::Call {
+                location: assign_location,
+                fun: value,
+                arguments: vec![CallArg {
+                    location: assign_location,
+                    label: None,
+                    value: UntypedExpr::lambda(name, continuation, assign_location),
+                }],
+            },
+        }
+    }
+
     fn infer_seq(&mut self, location: Span, untyped: Vec<UntypedExpr>) -> Result<TypedExpr, Error> {
+        // Search for backpassing.
+        let mut breakpoint = None;
+        let mut prefix = Vec::with_capacity(untyped.len());
+        let mut suffix = Vec::with_capacity(untyped.len());
+        for expression in untyped.into_iter() {
+            match expression {
+                _ if breakpoint.is_some() => suffix.push(expression),
+                UntypedExpr::Assignment {
+                    kind: AssignmentKind::Bind,
+                    ..
+                } => {
+                    breakpoint = Some(expression);
+                }
+                _ => prefix.push(expression),
+            }
+        }
+        if let Some(breakpoint) = breakpoint {
+            prefix.push(self.backpass(breakpoint, suffix));
+            return self.infer_seq(location, prefix);
+        }
```

Comment:
Note: I have been carefully avoiding cloning and re-allocations here. We do still traverse sequence twice now though, but the first pass should be pretty fast as it doesn't re-allocate anything and just look for Assignment expressions. 

If we find one, we rewrite the sequence and re-infer from that new sequence (in case of nested backpassing). Which should work pretty efficiently.